spring-integration-core/src/main/java/org/springframework/integration/channel/interceptor/ThreadStatePropagationChannelInterceptor.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.integration.channel.interceptor;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.ChannelInterceptorAdapter;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+
+/**
+ * The {@link ExecutorChannelInterceptor} implementation responsible for
+ * the {@link Thread} (any?) state propagation from one message flow's thread to another
+ * through the {@link MessageChannel}s involved in the flow.
+ * <p>
+ * The propagation is done from the {@link #preSend(Message, MessageChannel)}
+ * implementation using some internal {@link Message} extension which keeps the message
+ * to send and the state to propagate.
+ * <p>
+ * The propagated state context extraction and population is from the {@link #postReceive}
+ * implementation for the {@link org.springframework.messaging.PollableChannel}s, and from
+ * {@link #beforeHandle} for the
+ * {@link org.springframework.integration.channel.AbstractExecutorChannel}s and
+ * {@link org.springframework.messaging.support.ExecutorSubscribableChannel}s
+ *
+ * @param <S> the propagated state object type.
+ *
+ * @author Artem Bilan
+ * @since 4.2
+ */
+public abstract class ThreadStatePropagationChannelInterceptor<S> extends ChannelInterceptorAdapter
+		implements ExecutorChannelInterceptor {
+
+	@Override
+	public final Message<?> preSend(Message<?> message, MessageChannel channel) {
+		S threadContext = obtainPropagatingContext(message, channel);
+		if (threadContext != null) {
+			return new MessageWithThreadState<S>(message, threadContext);
```

Comment:
We will have to be careful to document that only one of these interceptors are allowed and it must be the last interceptor in the list. Otherwise a subsequent interceptor might remove the wrapper.